File: server/browse_web.js
#!/usr/bin/env node
/**
 * GIA Web Browser Agent — navigate, fill forms, click, screenshot, extract.
 * Uses Playwright with system Chromium.
 * Input: JSON as command-line argument or stdin {action, url, selector, value, query, options}
 * Actions: navigate, extract, fill, click, screenshot, search, get_text, get_html, scroll
 */
const { chromium } = require('playwright');

const CHROMIUM_PATH = '/usr/bin/chromium-browser';

const ACTIONS = {
  navigate: async (page, args) => {
    await page.goto(args.url, { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(args.options?.wait ?? 1000);
    return {
      url: page.url(),
      title: await page.title(),
      content: await page.evaluate(() => document.body.innerText.slice(0, 10000)),
    };
  },

  extract: async (page, args) => {
    const sel = args.selector;
    const timeout = args.options?.timeout ?? 5000;
    if (!sel) {
      return { text: await page.evaluate(() => document.body.innerText.slice(0, 20000)) };
    }
    try {
      await page.waitForSelector(sel, { timeout });
      const text = await page.$eval(sel, el => el.textContent);
      const html = await page.$eval(sel, el => el.outerHTML);
      return { text: text?.trim(), html };
    } catch {
      return { error: `Selector '${sel}' not found` };
    }
  },

  fill: async (page, args) => {
    const { selector, value } = args;
    const timeout = args.options?.timeout ?? 5000;
    try {
      await page.waitForSelector(selector, { timeout });
      await page.click(selector);
      await page.fill(selector, '');
      await page.fill(selector, value);
      return { success: true, filled: selector };
    } catch (e) {
      return { error: `Failed to fill '${selector}': ${e.message}` };
    }
  },

  click: async (page, args) => {
    const { selector } = args;
    const timeout = args.options?.timeout ?? 5000;
    try {
      await page.waitForSelector(selector, { timeout });
      await page.click(selector);
      await page.waitForTimeout(args.options?.wait ?? 1000);
      return { success: true, clicked: selector, url: page.url(), title: await page.title() };
    } catch (e) {
      return { error: `Failed to click '${selector}': ${e.message}` };
    }
  },

  screenshot: async (page, args) => {
    const buf = await page.screenshot({ type: 'png', fullPage: args.options?.fullPage ?? false });
    return { screenshot: buf.toString('base64'), format: 'png', size: buf.length };
  },

  get_html: async (page, args) => {
    if (args.selector) {
      const htmls = await page.$$eval(args.selector, els => els.map(el => el.outerHTML));
      return { html: htmls };
    }
    return { html: await page.content() };
  },

  get_text: async (page) => {
    return { text: await page.evaluate(() => document.body.innerText.slice(0, 20000)) };
  },

  scroll: async (page, args) => {
    const dir = args.direction ?? 'down';
    const amt = args.amount ?? 500;
    switch (dir) {
      case 'down': await page.evaluate(y => window.scrollBy(0, y), amt); break;
      case 'up': await page.evaluate(y => window.scrollBy(0, -y), amt); break;
      case 'bottom': await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight)); break;
      case 'top': await page.evaluate(() => window.scrollTo(0, 0)); break;
    }
    return { scrolled: dir, pixels: amt };
  },
};

async function main() {
  const raw = process.argv[2] || await new Promise(r => {
    let d = '';
    process.stdin.on('data', c => d += c);
    process.stdin.on('end', () => r(d));
  });

  let args;
  try { args = JSON.parse(raw); } catch { args = { action: 'navigate', url: raw }; }

  const action = args.action || 'navigate';
  const handler = ACTIONS[action];
  if (!handler) throw new Error(`Unknown action: ${action}. Valid: ${Object.keys(ACTIONS).join(', ')}`);

  const browser = await chromium.launch({
    headless: true,
    executablePath: CHROMIUM_PATH,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage', '--disable-gpu'],
  });

  const context = await browser.newContext({
    viewport: { width: 1280, height: 720 },
    userAgent: 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36 GIA/1.0',
  });
  const page = await context.newPage();

  try {
    if (args.url && !['search'].includes(action)) {
      await page.goto(args.url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    }
    const result = await handler(page, args);
    result.success = true;
    result.action = action;
    result.url = page.url();
    result.title = await page.title();
    console.log(JSON.stringify(result));
  } catch (e) {
    console.log(JSON.stringify({ success: false, error: e.message, action }));
  } finally {
    await browser.close();
  }
}

main();

File: server/browse_web.py
#!/usr/bin/env python3
"""GIA HTTP Web Browser — reliable web fetching, parsing, form submission.
Uses requests + BeautifulSoup. Works in PRoot without browser binary.
Input: JSON {action, url, data, headers, selector, options}
Actions: get, post, extract, search, submit_form
"""
import sys, json, re
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlencode, quote_plus

def main():
    if len(sys.argv) > 1:
        raw = sys.argv[1]
    else:
        raw = sys.stdin.read().strip()
    try:
        args = json.loads(raw)
    except:
        args = {"action": "get", "url": raw}

    action = args.get("action", "get")
    url = args.get("url", "")
    data = args.get("data", {})
    headers = args.get("headers", {"User-Agent": "Mozilla/5.0 GIA/1.0"})
    selector = args.get("selector", "")
    options = args.get("options", {})
    timeout = options.get("timeout", 20)

    try:
        if action == "get":
            r = requests.get(url, headers=headers, timeout=timeout)
            r.raise_for_status()
            soup = BeautifulSoup(r.text, "lxml")
            result = {"url": r.url, "status": r.status_code, "title": soup.title.string if soup.title else "", "content_type": r.headers.get("content-type", "")}
            if selector:
                els = soup.select(selector)
                result["elements"] = [str(el) for el in els]
                result["text"] = "\n".join(el.get_text(strip=True) for el in els)
            else:
                for tag in soup(["script", "style", "nav", "footer", "header", "noscript"]):
                    tag.decompose()
                result["text"] = soup.get_text(separator="\n").strip()[:30000]
            print(json.dumps({**result, "success": True}))

        elif action == "post":
            r = requests.post(url, data=data, headers=headers, timeout=timeout)
            r.raise_for_status()
            print(json.dumps({"success": True, "url": r.url, "status": r.status_code, "text": r.text[:10000]}))

        elif action == "search":
            query = args.get("query", "")
            r = requests.get(f"https://duckduckgo.com/html/?q={quote_plus(query)}", headers=headers, timeout=timeout)
            soup = BeautifulSoup(r.text, "lxml")
            results = []
            for a in soup.select("a.result__snippet, a.result__url"):
                pass
            for res in soup.select(".result"):
                t = res.select_one(".result__snippet, .result__title")
                u = res.select_one("a.result__snippet, a.result__url, a[href]")
                if t and u:
                    results.append({"title": t.get_text(strip=True), "url": u.get("href", ""), "snippet": t.get_text(strip=True)})
            print(json.dumps({"success": True, "query": query, "results": results[:10], "count": len(results)}))

        elif action == "extract":
            r = requests.get(url, headers=headers, timeout=timeout)
            soup = BeautifulSoup(r.text, "lxml")
            if selector:
                els = soup.select(selector)
                print(json.dumps({"success": True, "elements": [str(e) for e in els], "text": "\n".join(e.get_text(strip=True) for e in els)}))
            else:
                for tag in soup(["script", "style"]): tag.decompose()
                print(json.dumps({"success": True, "text": soup.get_text(separator="\n").strip()[:30000]}))

        elif action == "submit_form":
            r = requests.get(url, headers=headers, timeout=timeout)
            soup = BeautifulSoup(r.text, "lxml")
            form = soup.select_one(selector) if selector else soup.find("form")
            if not form:
                print(json.dumps({"success": False, "error": "No form found"})); return
            action_url = form.get("action") or url
            method = form.get("method", "get").lower()
            inputs = {inp.get("name"): inp.get("value", "") for inp in form.find_all(["input", "select", "textarea"]) if inp.get("name")}
            inputs.update(data)
            if method == "post":
                r = requests.post(urljoin(url, action_url), data=inputs, headers=headers, timeout=timeout)
            else:
                r = requests.get(urljoin(url, action_url), params=inputs, headers=headers, timeout=timeout)
            r.raise_for_status()
            print(json.dumps({"success": True, "url": r.url, "status": r.status_code, "text": r.text[:10000]}))

        else:
            print(json.dumps({"success": False, "error": f"Unknown action: {action}"}))

    except Exception as e:
        print(json.dumps({"success": False, "error": str(e), "action": action}))

if __name__ == "__main__":
    main()